### PR TITLE
change embed keyword to reference keyword

### PR DESCRIPTION
**Pull request by itkg-nanne**
[OO-DEPRECATED] deprecated OpenOrchestra\Backoffice\Form\DataTransformer\EmbedKeywordsToKeywordsTransformer
[OO-FEATURE] stock reference keyword

https://github.com/open-orchestra/open-orchestra-cms-bundle/pull/1662
https://github.com/open-orchestra/open-orchestra-media-admin-bundle/pull/232
https://github.com/open-orchestra/open-orchestra-model-bundle/pull/583
https://github.com/open-orchestra/open-orchestra-model-interface/pull/188
https://github.com/open-orchestra/open-orchestra-mongo-libs/pull/31
https://github.com/open-orchestra/open-orchestra-libs/pull/51
https://github.com/open-orchestra/open-orchestra-display-bundle/pull/219
https://github.com/open-orchestra/open-orchestra-media-bundle/pull/201
https://github.com/open-orchestra/open-orchestra-elastica-bundle/pull/35

### CODE REVIEW

File: Backoffice/Form/DataTransformer/EmbedKeywordsToKeywordsTransformer.php [hunk 1]
```diff
@@ -8,6 +8,7 @@
 use Symfony\Component\Form\DataTransformerInterface;
 use Doctrine\Common\Collections\ArrayCollection;
 use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+@trigger_error('The '.__NAMESPACE__.'\EmbedKeywordsToKeywordsTransformer class is deprecated since version 1.2.0 and will be removed in 1.2.1', E_USER_DEPRECATED);
```
Comment (alavieille) [inline Backoffice/Form/DataTransformer/EmbedKeywordsToKeywordsTransformer.php]: removed in 1.3.0, add deprecated annotation in php doc

File: Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+use OpenOrchestra\ModelInterface\Form\DataTransformer\ConditionFromBooleanToBddTransformerInterface;
+
+/**
+ * Class ConditionToReferenceKeywordTransformer
+ */
+class ConditionToReferenceKeywordTransformer implements DataTransformerInterface
+{
+    protected $keywordToDocumentManager;
+    protected $keywordRepository;
+
```
Comment (alavieille) [inline Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php]: remove this empty line

File: Backoffice/Manager/KeywordToDocumentManager.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+use OpenOrchestra\Backoffice\NavigationPanel\Strategies\AdministrationPanelStrategy;
+use OpenOrchestra\ModelInterface\Model\KeywordInterface;
+
+
+/**
+ * Class KeywordToDocumentManager
+ */
+class KeywordToDocumentManager
+{
+
```
Comment (alavieille) [inline Backoffice/Manager/KeywordToDocumentManager.php]: remove this empty line

File: Backoffice/Manager/KeywordToDocumentManager.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace OpenOrchestra\Backoffice\Manager;
+
```
Comment (alavieille) [inline Backoffice/Manager/KeywordToDocumentManager.php]: remove this empty line

File: Backoffice/Tests/Form/DataTransformer/ConditionToReferenceKeywordTransformerTest.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        Phake::when($this->keywordToDocumentManager)->getDocument('cat:X2')->thenReturn($catX2Keyword);
+        Phake::when($this->keywordToDocumentManager)->getDocument('author:AAA')->thenReturn($authorAAAKeyword);
+        Phake::when($this->keywordToDocumentManager)->getDocument('T1')->thenReturn($t1Keyword);
+        Phake::when($this->keywordToDocumentManager)->getDocument('T2')->thenReturn($t2Keyword);
+        Phake::when($this->keywordToDocumentManager)->getDocument('T3')->thenReturn($t3Keyword);
+        Phake::when($this->keywordToDocumentManager)->getDocument('not_created_keyword')->thenReturn($notCreatedKeyword);
+
+        $this->transformer = new ConditionToReferenceKeywordTransformer($this->keywordToDocumentManager, $this->keywordRepository);
+    }
+    /**
```
Comment (alavieille) [inline Backoffice/Tests/Form/DataTransformer/ConditionToReferenceKeywordTransformerTest.php]: add an empty line

File: Backoffice/Tests/Form/DataTransformer/ConditionToReferenceKeywordTransformerTest.php [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+     * @param string $string
+     *
+     * @dataProvider providerReverseTransformData
+     */
+    public function testReverseTransform($string, $expected)
+    {
+        $this->assertSame($expected, $this->transformer->reverseTransform($string));
+    }
+    /**
```
Comment (alavieille) [inline Backoffice/Tests/Form/DataTransformer/ConditionToReferenceKeywordTransformerTest.php]: add an empty line

File: Backoffice/Tests/Form/DataTransformer/ConditionToReferenceKeywordTransformerTest.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+     * @param string $string
+     *
+     * @dataProvider providerTransformData
+     */
+    public function testTransform($string, $expected)
+    {
+        $this->assertSame($expected, $this->transformer->transform($string));
+    }
+    /**
```
Comment (alavieille) [inline Backoffice/Tests/Form/DataTransformer/ConditionToReferenceKeywordTransformerTest.php]: add an empty line

File: Backoffice/Tests/Form/DataTransformer/CsvToReferenceKeywordTransformerTest.php [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        $this->keywordToDocumentManager = Phake::mock('OpenOrchestra\Backoffice\Manager\KeywordToDocumentManager');
+        $this->keywordRepository = Phake::mock('OpenOrchestra\ModelInterface\Repository\KeywordRepositoryInterface');
+
+        $this->keywordRepository = Phake::mock('OpenOrchestra\ModelInterface\Repository\KeywordRepositoryInterface');
+
+        $this->keywordToDocumentManager = Phake::mock('OpenOrchestra\Backoffice\Manager\KeywordToDocumentManager');
+
+        $this->transformer = new CsvToReferenceKeywordTransformer($this->keywordToDocumentManager, $this->keywordRepository);
+    }
+    /**
```
Comment (alavieille) [inline Backoffice/Tests/Form/DataTransformer/CsvToReferenceKeywordTransformerTest.php]: add an empty line

File: Backoffice/Tests/Form/DataTransformer/CsvToReferenceKeywordTransformerTest.php [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+     * @param array  $keywords
+     *
+     * @dataProvider providerReverseTransformData
+     */
+    public function testReverseTransform($string, $keywords)
+    {
+        $keywords = $this->createKeywordsToInterface($keywords);
+        $this->assertSame($keywords, $this->transformer->reverseTransform($string)->toArray());
+    }
+    /**
```
Comment (alavieille) [inline Backoffice/Tests/Form/DataTransformer/CsvToReferenceKeywordTransformerTest.php]: add an empty lien

File: Backoffice/Tests/Manager/KeywordToDocumentManagerTest.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+ * Class KeywordToDocumentManagerTest
+ */
+class KeywordToDocumentManagerTest extends AbstractBaseTestCase
+{
+    protected $manager;
+    protected $keywordRepository;
+    protected $suppressSpecialCharacterHelper;
+    protected $keywordClass;
+    protected $authorizationChecker;
+
```
Comment (alavieille) [inline Backoffice/Tests/Manager/KeywordToDocumentManagerTest.php]: remove this empty line

File: Backoffice/Validator/Constraints/BooleanConditionValidator.php [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+use Symfony\Component\Validator\ConstraintValidator;
+use OpenOrchestra\ModelInterface\Form\DataTransformer\ConditionFromBooleanToBddTransformerInterface;
+
+/**
+ * Class BooleanConditionValidator
+ */
+class BooleanConditionValidator extends ConstraintValidator
+{
+    /**
+     * @param string $value
```
Comment (alavieille) [inline Backoffice/Validator/Constraints/BooleanConditionValidator.php]: align phpdoc

File: Backoffice/Validator/Constraints/BooleanConditionValidator.php [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    protected function validateCondition($condition)
+    {
+        $is_boolean = true;
+        $encapsuledElements = array();
+        preg_match_all(
+            ConditionFromBooleanToBddTransformerInterface::GET_BALANCED_BRACKETS,
+            $condition,
+            $encapsuledElements
+        );
+        if(count($encapsuledElements[0]) > 0) {
```
Comment (alavieille) [inline Backoffice/Validator/Constraints/BooleanConditionValidator.php]: what happens if index 0 not exist ?

File: BackofficeBundle/Resources/config/transformer.yml [hunk 13]
```diff
@@ -1,22 +1,15 @@
...
     open_orchestra_backoffice.transformer.choice_string_to_array.class: OpenOrchestra\Backoffice\Form\DataTransformer\ChoiceStringToArrayTransformer
     open_orchestra_backoffice.transformer.bbcode_to_html.class: OpenOrchestra\Backoffice\Form\DataTransformer\BBcodeToHtmlTransformer
     open_orchestra_backoffice.transformer.reference_to_embed.class: OpenOrchestra\Backoffice\Form\DataTransformer\ReferenceToEmbedTransformer
     open_orchestra_backoffice.transformer.block_to_array.class: OpenOrchestra\Backoffice\Form\DataTransformer\BlockToArrayTransformer
-
+    open_orchestra_backoffice.transformer.csv_to_reference_keyword.class: OpenOrchestra\Backoffice\Form\DataTransformer\CsvToReferenceKeywordTransformer
+    open_orchestra_backoffice.transformer.condition_to_reference_keyword.class: OpenOrchestra\Backoffice\Form\DataTransformer\ConditionToReferenceKeywordTransformer
+    
 services:
-    open_orchestra_backoffice.transformer.embedkeywords_to_keywords:
```
Comment (alavieille) [inline BackofficeBundle/Resources/config/transformer.yml]: don't remove this service, can you mark it like deprecated

File: Backoffice/Validator/Constraints/BooleanConditionValidator.php (hunk 12 above)
Comment (itkg-nanne) [inline Backoffice/Validator/Constraints/BooleanConditionValidator.php]: could not happen : preg_match_all will return at last an empty array for index 0 if pattern is correctly coded. Add it in test.

File: BackofficeBundle/Resources/config/transformer.yml (hunk 13 above)
Comment (itkg-nanne) [inline BackofficeBundle/Resources/config/transformer.yml]: @alavieille , open_orchestra_model.document.embed_keyword.class don't exist anymore. Should i add it too?

**Comment by itkg-nanne** [comment]
update done @alavieille

File: Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+     * @return string
+     */
+    public function transform($keywords)
+    {
+        if (null === $keywords) {
+            return '';
+        }
+        $matches = array();
+        preg_match_all('/' . ConditionFromBooleanToBddTransformerInterface::SEPARATOR . '(.*?)' . ConditionFromBooleanToBddTransformerInterface::SEPARATOR . '/', $keywords, $matches);
+        foreach ($matches[1] as $id) {
```
Comment (alavieille) [inline Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php]: may be test if $matches[1] exist

File: Backoffice/Tests/Validator/Constraints/BooleanConditionValidatorTest.php [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+class BooleanConditionValidatorTest extends AbstractBaseTestCase
+{
+    /**
+     * @var BooleanConditionValidator
+     */
+    protected $validator;
+    protected $context;
+    protected $constraint;
+    protected $constraintViolationBuilder;
+
```
Comment (alavieille) [inline Backoffice/Tests/Validator/Constraints/BooleanConditionValidatorTest.php]: remove this empty line

File: Backoffice/Validator/Constraints/BooleanConditionValidator.php (hunk 12 above)
Comment (alavieille) [inline Backoffice/Validator/Constraints/BooleanConditionValidator.php]: @itkg-canne preg_match_all can return false if there is an error

Comment (itkg-nanne) [inline Backoffice/Validator/Constraints/BooleanConditionValidator.php]: @alavieille you're speaking about preg_match_all return but encapsuledElements is not preg_match_all return.

File: Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * @param KeywordToDocumentManager   $keywordToDocumentManager
+     * @param KeywordRepositoryInterface $keywordRepository
+     */
+    public function __construct(
+        KeywordToDocumentManager $keywordToDocumentManager,
+        KeywordRepositoryInterface $keywordRepository
+    )
+    {
```
Comment (itkg-ngilain) [inline Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php]: @itkg-nanne to be psr, this should be ``){`` on the same line

File: Backoffice/Manager/KeywordToDocumentManager.php [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        $keyword = $this->suppressSpecialCharacterHelper->transform($keyword);
+        $keywordClass = $this->keywordClass;
+        $keywordEntity = $this->keywordRepository->findOneByLabel($keyword);
+        if (is_null($keywordEntity) && $this->authorizationChecker->isGranted(AdministrationPanelStrategy::ROLE_ACCESS_CREATE_KEYWORD)) {
+            $keywordEntity = new $keywordClass();
+            $keywordEntity->setLabel($keyword);
+            $this->keywordRepository->getManager()->persist($keywordEntity);
+            $this->keywordRepository->getManager()->flush($keywordEntity);
+        }
```
Comment (itkg-ngilain) [inline Backoffice/Manager/KeywordToDocumentManager.php]: @itkg-nanne I think à 403 exception should be throwed somewhere

File: Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        foreach ($keywordArray as $keyword) {
+            if ($keyword != '') {
+                $keywordDocument = $this->keywordToDocumentManager->getDocument($keyword);
+                if($keywordDocument instanceof KeywordInterface) {
+                    $keywords = str_replace($keyword, $keywordDocument->getId(), $keywords);
+                } else {
+                    throw new NotFoundedKeywordException();
+                }
+            }
+        }
```
Comment (itkg-ngilain) [inline Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php]: this code is mainly the same as ``tranform`` You should create a protected function injecting it the specific variables

File: Backoffice/Form/DataTransformer/CsvToReferenceKeywordTransformer.php [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+{
+    protected $keywordToDocumentManager;
+
+    /**
+     * @param KeywordToDocumentManager $keywordToDocumentManager
+     */
+    public function __construct(
+        KeywordToDocumentManager $keywordToDocumentManager
+    )
+    {
```
Comment (itkg-ngilain) [inline Backoffice/Form/DataTransformer/CsvToReferenceKeywordTransformer.php]: ```){``

File: Backoffice/Form/Type/Component/KeywordsChoiceType.php [hunk 20]
```diff
@@ -7,66 +7,60 @@
...
      */
     public function __construct(
-        EmbedKeywordsToKeywordsTransformer $keywordsTransformer,
+        CsvToReferenceKeywordTransformer $csvToReferenceKeywordTransformer,
+        ConditionToReferenceKeywordTransformer $conditionToReferenceKeywordTransformer,
         KeywordRepositoryInterface $keywordRepository,
         RouterInterface $router,
         AuthorizationCheckerInterface $authorizationChecker
     )
     {
```
Comment (itkg-ngilain) [inline Backoffice/Form/Type/Component/KeywordsChoiceType.php]: ){

File: Backoffice/Manager/KeywordToDocumentManager.php [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+     * @param string                                  $keywordClass
+     * @param AuthorizationCheckerInterface           $authorizationChecker
+     */
+    public function __construct(
+        KeywordRepositoryInterface $keywordRepository,
+        SuppressSpecialCharacterHelperInterface $suppressSpecialCharacterHelper,
+        $keywordClass,
+        AuthorizationCheckerInterface $authorizationChecker
+    )
+    {
```
Comment (itkg-ngilain) [inline Backoffice/Manager/KeywordToDocumentManager.php]: ){

File: Backoffice/Validator/Constraints/BooleanConditionValidator.php [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    protected function validateCondition($condition)
+    {
+        $is_boolean = true;
+        $encapsuledElements = array();
+        preg_match_all(
+            ContentRepositoryInterface::GET_BALANCED_BRACKETS,
+            $condition,
+            $encapsuledElements
+        );
+        if(count($encapsuledElements[0]) > 0) {
```
Comment (itkg-ngilain) [inline Backoffice/Validator/Constraints/BooleanConditionValidator.php]: if (

**Comment by itkg-nanne** [comment]
update done @itkg-ngilain

File: Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php (hunk 18 above)
Comment (itkg-ngilain) [inline Backoffice/Form/DataTransformer/ConditionToReferenceKeywordTransformer.php]: The comment was mainly for the big foreach

**Comment by itkg-ngilain** [comment]
:+1:

**Pull request closed without merging**